tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,338 @@
+        only once and then passed on.
+
+        Currently, this optimizer does not preserve optimizer state. Its state preservation methods will
+        differ significantly from a standard optimizer because it is a wrapper for multiple optimizers each with
+        their own learning rate, hyper parameters, and slots.
+
+        Example usage
+            model = tf.keras.Sequential()
+            model.add(tf.keras.applications.resnet.ResNet50(include_top = False, pooling = 'avg'))
+            model.add(tf.keras.layers.Dense(1, activation = 'sigmoid'))
+            model.layers[0].lr_mult = 0.01
+            opt = DiscriminativeWrapper(tf.keras.optimizers.Adam, model, learning_rate = 0.01)
+            model.compile(loss = tf.keras.losses.BinaryCrossentropy, optimizer = opt)
+            model.fit(x, y)
+
+        Arguments
+            base_optimizer: A class that inherits from tf.keras.optimizers.Optimizer. Do not
+                pass an instance of the class.
+
+            model: tf.keras.Model, The model to be used for discriminative learning.
+                It should have at least 1 layer with the attribute lr_mult. The lr_mult should
+                be set to a value not equal to 1. Otherwise, you will have the exact same
+                result as not using discriminative learning.
+
+            learning_rate: float, the learning rate for the model
+
+            verbose: Bool, to generate a report on how many parameters are affected
+
+            *args: Args to pass to the base optimizer
+
+            **kwargs: Kwargs to pass to the base optimizer
+
+        Returns
+            Optimizer - A keras optimizer to use with model.compile
+
+        References
+            - [Universal Language Model Fine-tuning for Text Classification](https://arxiv.org/pdf/1801.06146.pdf)
+        """
+
+        assert issubclass(
+            base_optimizer, tf.keras.optimizers.Optimizer
+        ), "Base optimizer must be a class that inherits from tf.keras.optimizers.Optimizer"
+
+        # assume that users will follow the general guidelines and init thier opts within a dist scope.
+        if tf.distribute.has_strategy():
+            logging.warning(
+                """The discriminative layer optimizer may not behave as expected
+            when using a distribution strategy.
+            """
+            )
+
+        super().__init__(lr=learning_rate, name=name, *args, **kwargs)
+
+        DiscriminativeModelManager._prepare_model(model, verbose=verbose)
+
+        self.opt_class = base_optimizer
+        self.learning_rate = learning_rate
+        self.kwargs = kwargs
+
+        # Find unique lr_mult.
+        variable_groups = {var.lr_mult: None for var in model.trainable_variables}
+
+        self.optimizer_group = []
+
+        for lr_mult in variable_groups.keys():
+            opt = self.opt_class(learning_rate=learning_rate * lr_mult, **kwargs)
+            opt.lr_mult = lr_mult
+            self.optimizer_group.append(opt)
```

Comment:
I believe a cleaner option would be here  to have only a single optimizer, and when doing `apply_gradiens`, apply the multiplier to the gradiens before giving them to the base optimizer. Similar to this: https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer#example_6 

So the user should give an optimizer in DiscriminativeLayerOptimizer, not the class of an optimizer. I believe that we'll also have way less serialization problems with this method.

Reply:
I tried that in an earlier version and it was the preferred method. However, it only works for SGD. 

This is because additional variables are stored for Adam and other momentum based optimizers, which do not interact identically for gradients and learning rates. As in, half the gradient is not equivalent to half the learning rate for an Adam optimizer.

Reply:
Earlier, I reviewed the fast AI implementation of discriminative layer training and I believe they used distinct sets of non-model variables for each learning rate, effectively the same as distinct optimizers for each learning rate.. 

These non-model variables interact with learning rates and gradients very differently compared to model variables. I believe this is because non-model variables use the gradient for complex calculations, such as first and second order moments. Passing a gradient multiplied by a scalar does not translate into a model variable delta multiplied by that scalar. So when affected the gradient instead of the learning rate of the optimizer, you will get different results. 

I have tested this using Adam at 0.01 LR and Adam at 0.02 LR but with the gradient multiplied by 0.5. The model variable deltas are different, which suggests that the non model variables are not scaling linearly. 

For SGD this works fine, because SGD multiplies the gradient by the learning rate to get the model variable delta.

Reply:
Could you give a link to the fastai implementation?

Reply:
This is their documentation
https://docs.fast.ai/basic_train.html#Discriminative-layer-training

The source is here in the optimwrapper. 
https://github.com/fastai/fastai/blob/9a32a5f666d28283184e4b5b30edebfcb8ddb5f3/fastai/callback.py#L9

Specifically, the create method appears to create a list of layer groups and learning rates, which the optimizer function interacts with.